haystack/nodes/sampler/top_p_sampler.py

```diff
@@ -0,0 +1,161 @@
+    the documents based on the cumulative probability of these scores. The TopPSampler provides a
+    way to efficiently select the most relevant documents based on their similarity to a given query.
+
+    Usage example:
+
+    ```python
+    prompt_node = PromptNode(
+        "text-davinci-003",
+        api_key=openai_key,
+        max_length=256,
+        default_prompt_template="question-answering-with-document-scores",
+    )
+    retriever = WebRetriever(api_key=os.environ.get("SERPERDEV_API_KEY"), mode="preprocessed_documents")
+    sampler = TopPSampler(top_p=0.95)
+
+    pipeline = WebQAPipeline(retriever=retriever, prompt_node=prompt_node, sampler=sampler)
+    print(pipeline.run(query="What's the secret of the Universe?"))
+    ```
+    """
+
+    def __init__(
+        self,
+        model_name_or_path: Union[str, Path] = "cross-encoder/ms-marco-MiniLM-L-6-v2",
+        top_p: Optional[float] = 1.0,
+        strict: Optional[bool] = False,
+        score_field: Optional[str] = "score",
+        use_gpu: Optional[bool] = True,
+        devices: Optional[List[Union[str, torch.device]]] = None,
+    ):
+        """
+        Initialize a TopPSampler.
+
+        :param model_name_or_path: Path to a pretrained sentence-transformers model.
+        :param top_p: Cumulative probability threshold for filtering the documents (usually between 0.9 and 0.99).
+        :param strict: If `top_p` is set to a low value and sampler returned no documents, then setting `strict` to
+        `False` ensures at least one document is returned. If `strict` is set to `True`, then no documents are returned.
+        :param score_field: The name of the field that should be used to store the scores a document's meta data.
+        :param use_gpu: Whether to use GPU (if available). If no GPUs are available, it falls back on a CPU.
+        :param devices: List of torch devices (for example, cuda:0, cpu, mps) to limit inference to specific devices.
+        """
+        super().__init__()
+
+        self.top_p = top_p
+        self.score_field = score_field
+        self.strict = strict
+        self.devices, _ = initialize_device_settings(devices=devices, use_cuda=use_gpu, multi_gpu=True)
+        self.cross_encoder = CrossEncoder(model_name_or_path, device=str(self.devices[0]))
+
+    def predict(self, query: str, documents: List[Document], top_p: Optional[float] = None) -> List[Document]:
+        """
+        Returns a list of documents filtered using `top_p`, based on the similarity scores between the query and the
+        documents whose cumulative probability is less than or equal to `top_p`.
+
+        :param query: Query string.
+        :param documents: List of Document.
+        :param top_p: Cumulative probability threshold for filtering the documents. If not provided, the top_p value
+        set during TopPSampler initialization is used.
+        :return: List of Documents sorted by (desc.) similarity with the query.
```

Comment:
I still don't know what "desc." stands for here

Reply:
Sorry, it means descending. Sth like `List of Documents sorted by similarity with the query in descending order.`